staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -746,7 +747,33 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
 		}
 		return nil, readResourceVersion, "", nil
 	}
-	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, pred.MatcherIndex())
+	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, matcherIndex(ctx, pred))
+}
+
+func matcherIndex(ctx context.Context, pred storage.SelectionPredicate) []storage.MatchValue {
+	re, _ := request.RequestInfoFrom(ctx)
+	if re == nil || !re.IsResourceRequest || re.Subresource != "" || !whitelistResource(re.Resource) {
+		return pred.MatcherIndex()
+	}
+	// there multiple pattern for listing pods
+	// 1. list with namespace: /api/v1/namespace/ns/pods
+	// 2. list pods without namespace and with --field-selector=metadata.namespace=ns
+
+	// #1 case
+	if re.Namespace != "" {
+		if _, found := pred.Field.RequiresExactMatch("metadata.namespace"); !found {
+			pred.Field = fields.OneTermEqualSelector("metadata.namespace", re.Namespace)
+		}
+	}
+	// #2 case
+	// 2.1 pred.Field will not have "metadata.namespace" field if the request is like /api/v1/pods
+	// 2.2 pred.Field will have "metadata.namespace" field if the request is like /api/v1/pods?fieldSelector=metadata.namespace=ns
+	return pred.MatcherIndex()
+}
```

Comment:
I don't think all this logic belongs here. Instead I would rather switch `MatcherIndex()` function to take a context as a parameter and use it there [without modifying pred itself].